### PR TITLE
Create ibm-health.hpp to handle health roll up of LED's on IBM systems

### PR DESCRIPTION
**Pull request by lxwinspur**
The original health.hpp can not meet IBM's Health Rollup function,
so we need to add a new ibm-health.hpp to meet IBM's needs.

Signed-off-by: George Liu <liuxiwei@inspur.com>

### CODE REVIEW

File: redfish-core/lib/ibm-health.hpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
+/*
+// Copyright (c) 2019 Intel Corporation
```
Comment (gtmills) [inline redfish-core/lib/ibm-health.hpp:2]: Can you remove?

**Comment by gtmills** [comment]
> The original health.hpp can not meet IBM's Health Rollup function,
so we need to add a new ibm-health.hpp to meet IBM's needs.

Why doesn't health.hpp work? Can you explain how this is different?

File: redfish-core/lib/ibm-health.hpp (hunk 1 above)
Comment (lxwinspur) [reply redfish-core/lib/ibm-health.hpp:2]: Sure, will update

Comment (lxwinspur) [reply redfish-core/lib/ibm-health.hpp:2]: Sure, will update

**Comment by ChicagoDuan** [comment]
> Why doesn't health.hpp work? Can you explain how this is different?

In health.hpp, it use associations to determine health and HealthRollUp  status (https://github.com/openbmc/bmcweb/blob/master/redfish-core/lib/health.hpp#L191).

For example, when the PowerSupply OperationalStatus is false, we need to create a new association ( such as :
/xyz/openbmc_project/inventory/system/chassis/motherboard/powersupply0/critical ). Only the path ending with "critical" or "warning" will be detected by the getAllStatusAssociations() function in health.hpp, and then HealthRollup and Health will become "Critical".

In our system, there is no process to monitor the OperationalStatus and create the association when OperationalStatus becomes false. So it is not suitable for our system.

I saw the callback_manager process in the initial commit of redfish_health_rollup.md on gerrit: https://gerrit.openbmc-project.xyz/c/openbmc/docs/+/21380/7/designs/redfish_health_rollup.md#76 .
But it's Intel-specific. Intel uses callback_manager to monitor the OperationalStatus and create critical associations when OperationalStatus becomes false.

> Can you explain how this is different?

In the new file ibm-health.hpp,  we do not use critical association. We directly check the OperationalStatus. If OperationalStatus is false, then health is critical. HealthRollUp is the same way, directly check the operationalstatus of the path it contains.

**Comment by ChicagoDuan** [comment]
Hi @gtmills
Do you have any comments on the implementation of Health and HealthRollUp?
I have submitted the PowerSupply Health and HealthRollUp. It uses ibm-health.hpp to get Health and HealthRollUp status.

**Comment by lkammath** [comment]
> Hi @gtmills
> Do you have any comments on the implementation of Health and HealthRollUp?
> I have submitted the PowerSupply Health and HealthRollUp. It uses ibm-health.hpp to get Health and HealthRollUp status.

Hi @ChicagoDuan we will bring up this in the local design call and close it. 

@gtmills i was not yet onboard when we we were deciding these stuff (To not follow the existing design and have IBM specific design). Are you saying the renaming is not required and keep the original name as it is and go with the IBM approach with out renaming it? or is there some thing else we are missing here?

File: redfish-core/lib/power_subsystem.hpp [hunk 2]
```diff
@@ -23,6 +24,28 @@ inline void
         "/redfish/v1/Chassis/" + chassisID + "/PowerSubsystem";
     asyncResp->res.jsonValue["PowerSupplies"]["@odata.id"] =
         "/redfish/v1/Chassis/" + chassisID + "/PowerSubsystem/PowerSupplies";
+
+    constexpr const std::array<const char*, 1> inventoryForChassis = {
```
[the remark this answers is not in the document]
Comment (ChicagoDuan) [reply redfish-core/lib/power_subsystem.hpp:28]: > So we have to add this for every parent resource?

Yes, it is used in the same way as the existing frameworks.

File: redfish-core/lib/power_supply.hpp [hunk 3]
```diff
@@ -494,6 +462,12 @@ inline void requestRoutesPowerSupply(App& app)
                                     "/PowerSubsystem/PowerSupplies/" +
                                     powerSupplyID;
 
+                                auto health =
+                                    std::make_shared<ibmHealthPopulate>(
```
[the remark this answers is not in the document]
Comment (ChicagoDuan) [reply redfish-core/lib/power_supply.hpp:466]: > Will we add this to other resources like fans, assemblies, processors, memory?

Yes, if you agree use ibm-health.hpp. I will add this to other resources and push them.

**Comment by ChicagoDuan** [comment]
> > In the new file ibm-health.hpp, we do not use critical association. We directly check the OperationalStatus. If OperationalStatus is false, then health is critical. HealthRollUp is the same way, directly check the operationalstatus of the path it contains.
> 
> How do we determine how to roll up to the parent resource ?
> I.e. how do we roll up from power supplies to powersubsystem eventually chassis?

In the powersubsystem and chassis, we will add sub paths that need to be checked. Look at line 41 of power_subsystem.hpp

File: redfish-core/lib/power_subsystem.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChicagoDuan) [reply redfish-core/lib/power_subsystem.hpp:28]: > So we have to add this for every parent resource?

The existing framework also need add code to every parent resource:
https://github.com/openbmc/bmcweb/blob/master/redfish-core/lib/chassis.hpp#L243

**Comment by lxwinspur** [comment]
so sorry, misoperation.

**Pull request closed without merging**